lib/internal/queue.js

```diff
@@ -16,32 +14,68 @@ export default function queue(worker, concurrency, payload) {
     var _worker = wrapAsync(worker);
     var numRunning = 0;
     var workersList = [];
+    const events = {
+        error: [],
+        drain: [],
+        saturated: [],
+        unsaturated: [],
+        empty: []
+    }
+
+    function on (event, handler) {
+        events[event].push(handler)
```

Comment:
That's quite an interesting way of handling both promises and the current API. Is the promise based API useful for any of the events besides `drain`?

Reply:
It's most useful for `drain`, but i can see utility coming from `unsaturated`, and making `error` reject.  I'm also happy making the interface uniform for all the event-style methods.